docs/flow/config-language/expressions/types_and_values.md

```diff
@@ -6,3 +6,145 @@ weight: 100
 ---
 
 # Types and values
+
+## Types
+
+River uses the following types for its values:
+
+* `number`: Any numeric value, like `3` or `3.14`.
+* `string`: A sequence of Unicode characters representing text, like `"Hello, world!"`.
+* `bool`: A boolean value, either `true` or `false`.
+* `array`: A sequence of values, like `[1, 2, 3]`. Elements within the
+  list are indexed by whole numbers, starting with zero.
+* `object`: A group of values which are identified by named labels, like
+  `{ name = "John" }`.
+* `function`: A value representing a routine which can be executed with
+  arguments to compute another value, like `env("HOME")`. Functions take zero
+  or more arguments as input and always return a single value as output.
+* `null`: A type that has no value.
+
+### Special Types
+
+#### Secrets
+
+A `secret` is a special type of string which is never displayed to the user.
+`string` values may be assigned to an attribute expecting a `secret`, but never
+the inverse; it is not possible to convert a secret to a string or assign a
+secret to an attribute expecting a string.
+
+Secrets cannot be constructed by users, but are returned by components which
+deal with sensitive values.
+
+#### Capsules
+
+River has a special type called a `capsule`, which represents an internal type
+used by Grafana Agent for building telemetry pipelines. Capsule values can
+never be constructed by the user, but can be used in expressions and assigned
+to attributes. Capsules are represented to the user as
+`capsule("SOME_INTENRAL_NAME")`. An attribute expecting a capsule can only be
+given a capsule of the same internal type. The specific type of capsule
+expected is explicitly documented for any component which uses or exports them.
```

Comment:
I know this is something I wrote earlier, but I think I want to change how we communicate what capsules are to users to hide the internal implementation details a little bit. 

I've been thinking that we say that capsules are a _category_ of types which are internal to Grafana Agent, and that each capsule type has a unique name like `capsule("metrics.Receiver")`. 

Then we can say that capsule types are not inter-compatible, and that an attribute expecting a `capsule("metrics.Receiver")` can only be given a value of the `capsule("metrics.Receiver")` type.